### PR TITLE
Details block: add name attribute for exclusive accordion

### LINKED ISSUE #56969: Details block: Exclusive accordions
## What problem does this address?

> Safari 17.2 now supports the name attribute on the `<details>` element — which moves some commonly-used functionality from needing JavaScript to being baked into HTML. If you use the name attribute to give a series of details elements the same name, then only one in the series will be open at a time.

```HTML
<details name="foobar" open>
  <summary>Item 1</summary>
  <p>Since this first item has an open attribute, it’s open by default.</p> 
</details>
<details name="foobar">
   <summary>Item 2</summary>
  <p>When you clicked this item, any other open item will automatically close.</p>
</details>
<details name="foobar">
   <summary>Item 3</summary>
  <p>Now, developers don’t have to write any JavaScript to get this behavior.</p> 
</details>
```

>Any time a user clicks to open a second item, the first item automatically closes. This can make for a better user experience, although do make sure it’s right for your project. Think about users who might want or need to compare one item to another. Consider whether to not it will be annoying to have to tap over and over on each bit of content in order to read each one. This pattern can be frustrating when used inappropriately — even inaccessible.

>Here’s an example to compare the difference. Try it in a [browser with support](https://caniuse.com/mdn-html_elements_details_name).

https://codepen.io/jensimmons/pen/wvNPBZa

Source: https://webkit.org/blog/14787/webkit-features-in-safari-17-2/

## What is your proposed solution?

Add a name attribute to the detials block and a text field to edit it.

### LINKED ISSUE #62251: Add support for 'name' attribute of 'details' block.
## What problem does this address?
The _details_ block is currently lacking support for its _name_ attribute.

Spec.:
https://html.spec.whatwg.org/multipage/interactive-elements.html

The attribute is supported in all recent browsers (except Firefox, god knows why...):
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details

While the "accordion-like" functionality supplied by the attribute might be misleading, the details block should still aim to support all HTML attributes defined in the spec, see also https://github.com/WordPress/gutenberg/issues/21584#issuecomment-1910899417_.

## What is your proposed solution?
Add the "name" attribute to the block settings.

### PR DESCRIPTION
**Pull request by Soean**
## What?
Fixes #56969

>The name content attribute gives the name of the group of related [details](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-details-element) elements that the element is a member of. Opening one member of this group causes other members of the group to close. If the attribute is specified, its value must not be the empty string.

See HTML Living standard of WHATWG: https://html.spec.whatwg.org/multipage/interactive-elements.html#the-details-element

## How?
Adds a `name` attribute to the details block.

## Demo
https://github.com/WordPress/gutenberg/assets/695201/88b33e40-807b-4572-9c56-9fba3ca63d3e

## Testing Instructions
Add multiple details blocks .

**Supported browser**: https://caniuse.com/mdn-html_elements_details_name
Chrome 120, Edge 120, Safari 17.2, Firefox is not supporting this attribute yet.

Example content:
```HTML
<!-- wp:details {"name":"foobar"} -->
<details class="wp-block-details" name="foobar"><summary>This is a summary 1</summary><!-- wp:paragraph {"placeholder":"Type / to add a hidden block"} -->
<p>Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint obcaecat cupiditat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
<!-- /wp:paragraph --></details>
<!-- /wp:details -->

<!-- wp:details {"name":"foobar"} -->
<details class="wp-block-details" name="foobar"><summary>This is a summary 2</summary><!-- wp:paragraph {"placeholder":"Type / to add a hidden block"} -->
<p>Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint obcaecat cupiditat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
<!-- /wp:paragraph --></details>
<!-- /wp:details -->

<!-- wp:details {"name":"foobar"} -->
<details class="wp-block-details" name="foobar"><summary>This is a summary 3</summary><!-- wp:paragraph {"placeholder":"Type / to add a hidden block"} -->
<p>Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint obcaecat cupiditat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
<!-- /wp:paragraph --></details>
<!-- /wp:details -->
```

### CODE REVIEW

**Comment by fabiankaegy** [comment]
With this underlying feature it may be nice to in the future add an "Accordion" wrapper block that automatically handles the name generation & assignment for any descendant detail blocks intern so users don't have to manually tinker with the names where they can easily make typos etc.

And in an ideal world all the details elements would then also sync there styling similar to how the query loop post template does it. 

But these are all just future improvement ideas. Love the addition :) Will review the code in minute 👍

File: packages/block-library/src/details/edit.js [hunk 1]
```diff
@@ -56,6 +56,16 @@ function DetailsEdit( { attributes, setAttributes, clientId } ) {
 							} )
 						}
 					/>
+					<TextControl
+						label="Name"
```
Comment (fabiankaegy) [inline packages/block-library/src/details/edit.js]: Do we have any precedence for not translating a label when it refers to the name of a HTML attribute?

File: packages/block-library/src/details/save.js [hunk 2]
```diff
@@ -4,12 +4,12 @@
...
 
 export default function save( { attributes } ) {
-	const { showContent } = attributes;
+	const { name, showContent } = attributes;
 	const summary = attributes.summary ? attributes.summary : 'Details';
 	const blockProps = useBlockProps.save();
 
 	return (
-		<details { ...blockProps } open={ showContent }>
+		<details { ...blockProps } name={ name } open={ showContent }>
```
Comment (fabiankaegy) [inline packages/block-library/src/details/save.js]: Won't this cause deprecation issues? Because the `name` attribute is added with an empty string if the value is empty. 

I think either it should be omitted entirely, and then it also won't be an issue for deprecations, or we need to add a proper deprecation in place.

**Review by fabiankaegy — COMMENTED** [review]
I left two minor comments. It works as intended and is a great enhancement. 

Though I'm a little worried about the browser support at this point in time. 

Yes Safari TP shipped it yesterday but looking at the global usage we are at less than 1% https://caniuse.com/mdn-html_elements_details_name

I'm worried introducing the setting now will be confusing because for the large majority of users the setting will have no effect.

File: packages/block-library/src/details/edit.js (hunk 1 above)
Comment (Soean) [reply packages/block-library/src/details/edit.js]: Good catch, this should be translatable.

File: packages/block-library/src/details/save.js (hunk 2 above)
Comment (Soean) [reply packages/block-library/src/details/save.js]: I changed it to `name={ name || null }` so this is only rendered, if name has a value.

**Comment by richtabor** [comment]
> I'm worried introducing the setting now will be confusing because for the large majority of users the setting will have no effect.

I share the same concern. 

I'm also not sure most folks would understand what "name" is, even with help text. Maybe developers, but that's about it.

**Comment by fabiankaegy** [comment]
> I'm also not sure most folks would understand what "name" is, even with help text. Maybe developers, but that's about it.

@richtabor that is why I think the wrapper accordion block would be nice. Because it could abstract that away

**Comment by fabiankaegy** [comment]
Maybe it would be an option to move the setting to the advanced panel for now

**Comment by Soean** [comment]
> Yes Safari TP shipped it yesterday but looking at the global usage we are at less than 1% https://caniuse.com/mdn-html_elements_details_name

I don't think the 1% is correct yet, but I agree, we should wait a little longer before we merge it into the release. Maybe it will soon be supported by Firefox.

> Maybe it would be an option to move the setting to the advanced panel for now

You are right, this should be an advanced feature.

**Comment by hanneslsm** [comment]
From UX perspective I find the current name input field implementation counterintuitive. 
For me it feels like adding a css class where every block with a class gets the styling applied.
However, here it is reversed. Every details block with the name get's grouped together but only one can be selected then. It's more like a creating a toggle. 

I don't have a solution yet but my gut says this feels too unpolished yet. First thing that comes in my mind would be reusing the tags/category UX [(it's the Query Controls I think)](https://wordpress.github.io/gutenberg/?path=/docs/components-querycontrols--docs) instead of a plain text field to emphasize that the blocks are somehow "grouped together". 

Also, judging from the screenshot, the input field seems to not have the correct height.

**Comment by thomas-price** [comment]
Browser support is now 78% .. 1% to 78% in 4 months
Is there any possibility of this being included in time for WP 6.6?

**Comment by richtabor** [comment]
I think this is fine. Are there any objections to the actual functionality?

I'm fine with keeping it under Settings too—"Advanced" should be as consistent as possible in my view, especially as we continue to add controls across blocks, like overrides in synced patterns. "Title Attribute" and "Red Attribute" are in the Navigation Link blocks' Settings panel for example. 

I think this copy explains it a bit better: 

```
Enables multiple Details blocks with the same name attribute to be connected, with only one open at a time. 
```

**Comment by t-hamano** [comment]
The only problem is that Firefox doesn't support this attribute yet. As far as I can tell, there doesn't seem to be any progress on this issue.

- https://bugzilla.mozilla.org/show_bug.cgi?id=1856460
- https://github.com/mozilla/standards-positions/issues/831
- https://github.com/mozilla/standards-positions/issues/578

**Comment by sntran** [comment]
I would like to add my opinion if that's okay. It's fine for this not working in Firefox. The field is available in the Inspector. If you know what `name` does, then you know the limitation. If you don't know, then you won't input into it. Even if you do, it's just another attribute on the element.

**Comment by thomas-price** [comment]
> I would like to add my opinion if that's okay. It's fine for this not working in Firefox. The field is available in the Inspector. If you know what `name` does, then you know the limitation. If you don't know, then you won't input into it. Even if you do, it's just another attribute on the element.

It is supported by Firefox: https://developer.mozilla.org/en-US/blog/html-details-exclusive-accordions/

**Comment by t-hamano** [comment]
Yes, now that it is supported in all major browsers I think we can move forward with this PR.

**Comment by t-hamano** [comment]
@Soean Is it possible to proceed with this PR again?

**Comment by dabowman** [comment]
Just bumping this. It would be wonderful to have. Looks like it's now supported across current browser versions.

**Comment by t-hamano** [comment]
I've resolved the conflicts and made it work with the latest Gutenberg. Regarding the help text, I've adopted the suggestion in [this comment](https://github.com/WordPress/gutenberg/pull/56971#issuecomment-2067005404).

Any feedback is welcome!

image

**Comment by hanneslsm** [comment]
> I've resolved the conflicts and made it work with the latest Gutenberg. Regarding the help text, I've adopted the suggestion in [this comment](https://github.com/WordPress/gutenberg/pull/56971#issuecomment-2067005404).
> 
> Any feedback is welcome!
> 

Thank you som much @t-hamano !

Can't say anything code-wise but I tested this PR on playground in Chrome and Firefox and it works like a charm.
Is there anything we need to do in order to get merged? Maybe we can get it in the next version? :)

**Comment by t-hamano** [comment]
Update:

I've resolved the conflict again. I've also changed it to reference the HTML element via the source field to save the name attribute value.

@WordPress/gutenberg-core Now this PR is ready, and I'd be happy to receive any final feedback.

Summary of this PR:

- This PR makes [the `name` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDetailsElement/name) of the Details block configurable.
- By grouping multiple Details blocks by the name attribute, you can limit the number of Details blocks that can be opened to one.
- When this PR was first submitted, browser support was not sufficient. However, now all major browsers support this attribute ([Can I Use](https://caniuse.com/mdn-html_elements_details_name)).
- If the name attribute is not filled in, the saved HTML will not change. This means that there should be no need to add a block deprecation.

File: packages/block-library/src/details/edit.js [hunk 3]
```diff
@@ -90,11 +90,13 @@ function DetailsEdit( { attributes, setAttributes, clientId } ) {
...
-			<details
-				{ ...innerBlocksProps }
-				open={ hasSelection || showContent }
-			>
-				<summary onClick={ ( event ) => event.preventDefault() }>
+			<details { ...innerBlocksProps } open={ isOpen }>
+				<summary
+					onClick={ ( event ) => {
+						event.preventDefault();
+						setIsOpen( ! isOpen );
```
Comment (fabiankaegy) [inline packages/block-library/src/details/edit.js:97]: Won't this change here have the sideeffect that clicking into the Summary Rich Text element is going to open / close the details element?

Comment (t-hamano) [reply packages/block-library/src/details/edit.js:97]: This shouldn't cause any issues, since I haven't added the `name` attribute to the block in the editor.

However, it's worth discussing whether we should add the `name` attribute behavior in the editor as well, to be consistent with the frontend.

Comment (fabiankaegy) [reply packages/block-library/src/details/edit.js:97]: Ignore me, I seem to have commented on a disconnected HEAD.

**Review by carolinan — APPROVED** [review]
I tested it on Windows 11 without issues:
Chrome Version 132.0.6834.84 (Official Build) (64-bit)
Firefox 134.0.1 (64-bit)

I don't mind that the label for the field is "Name attribute", even if it is technical, the help text explains what the purpose is.

I don't think the word "attribute" will discourage people from using it. It can also be updated again if there is feedback from more users.

If browser support is still a concern, maybe the help text could be updated to include something along the lines of "in supported browsers".

**Review by fabiankaegy — APPROVED** [review]
Also in favor of merging this now :)

**Review by Mamaduka — APPROVED** [review]
Good work!

Do we want a dev note to showcase the feature?

**Comment by fabiankaegy** [comment]
> Good work!
> 
> Do we want a dev note to showcase the feature?

@Mamaduka given that it isn't really a change but rather a feature addition I don't think a Dev Note is needed. 

We should however feature it in the "Whats new in Gutenberg" post when it is released and given the scope of 6.8 I could see this being included in the highlight grid etc.

**Comment by t-hamano** [comment]
Thank you everyone for the reviews!

**Comment by swissspidy** [comment]
> > I'm worried introducing the setting now will be confusing because for the large majority of users the setting will have no effect.
> 
> I share the same concern.
> 
> I'm also not sure most folks would understand what "name" is, even with help text. Maybe developers, but that's about it.

Has anyone thought about this again? While I love the new support in general, I also find the current implementation confusing. It's absolutely not clear to users what this does and that it needs multiple details blocks to share the same name (a name which isn't even user-facing FWIW). You can't even preview the behavior in the editor.

A wrapping accordion block or something would be more intuitive.

**Comment by Mamaduka** [comment]
We could remove UI control and make it a developer-only feature, allowing folks to use this behavior in patterns, templates, etc. I think we have already done something similar for `ariaLabel` attributes.

> A wrapping accordion block or something would be more intuitive.

The idea makes sense; it is probably worth creating a separate discussion.

**Comment by hanneslsm** [comment]
> I'm also not sure most folks would understand what "name" is, even with help text. Maybe developers, but that's about it.

This is again the foundational discussion of who the target user of the editor is.
I argue that the "advanced" tab is for developers or at least users who are willing to read the docs.

> We could remove UI control and make it a developer-only feature, allowing folks to use this behavior in patterns, templates, etc. I think we have already done something similar for ariaLabel attributes.

I build custom themes for my clients, with patterns, templates, etc.. I present the websites to my clients, we go live, everyone is happy. Sometimes my client still wants small changes, like exactly this feature. Deploying a new version just for adding a name tag makes me shiver. This contradicts the whole purpose of a page builder and is the fundamental reason why Gutenberg is disliked by so many. 
in short: let's not remove the UI. We can think about a different UI/UX though. As said in [my comment above](https://github.com/WordPress/gutenberg/pull/56971#issuecomment-1851990286) a quick fix to improve the UX could be a autocomplete of existing name tags similar to tag/categories.

**Comment by t-hamano** [comment]
> I also find the current implementation confusing.

Could this be alleviated somewhat by improving the help text?

name

> let's not remove the UI. We can think about a different UI/UX though.

I agree with this direction.
